### PR TITLE
Add Applicative instance for Unfold

### PR DESCRIPTION
**Pull request by harendra-kumar**

### CODE REVIEW

**Comment by harendra-kumar** [comment]
```
Data.Unfold(cpuTime)
Benchmark                                                    default(μs)
------------------------------------------------------------ -----------
Data.Unfold/o-1-space/outer-product/toNullAp                       47.87
Data.Unfold/o-1-space/outer-product/toNull                         42.40
```

**Comment by harendra-kumar** [comment]
There was a bug in `Data.Unfold.const` which has been fixed in this PR.

File: src/Streamly/Internal/Data/Unfold/Types.hs [hunk 1]
```diff
@@ -63,3 +67,86 @@ map f (Unfold ustep uinject) = Unfold step uinject
...
+{-# INLINE const #-}
+const :: Applicative m => m b -> Unfold m a b
+const m = Unfold step inject
+
+    where
+
+    inject _ = pure False
+
+    step False = (`Yield` True) <$> m
+    step True = pure Stop
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Unfold/Types.hs:84]: This does not feel right to me. Discarding the `inject` value should be signified in the type.

```
const :: Applicative m => m b -> Unfold m Void b
```

Comment (pranaysashank) [reply src/Streamly/Internal/Data/Unfold/Types.hs:84]: `const` has to ignore it, you get that from the regular `const :: a -> b -> a`

Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Unfold/Types.hs:84]: We want to accept an input and discard it. I am not sure how that would be possible with `Void` input type.

File: src/Streamly/Internal/Data/Unfold/Types.hs [hunk 2]
```diff
@@ -63,3 +67,86 @@ map f (Unfold ustep uinject) = Unfold step uinject
...
+        s1 <- inject1 x
+        return $ CrossOuter s1 x
+
+    {-# INLINE_LATE step #-}
+    step (CrossOuter st1 sy) = do
+        r <- step1 st1
+        case r of
+            Yield x s -> do
+                s2 <- inject2 sy
+                return $ Skip (CrossInner s sy s2 x)
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Unfold/Types.hs:125]: You can probably inline `CrossInner`. Saves 1 indirection I guess.

File: src/Streamly/Internal/Data/Unfold/Types.hs [hunk 3]
```diff
@@ -63,3 +67,86 @@ map f (Unfold ustep uinject) = Unfold step uinject
...
+-- | Create a cross product (vector product or cartesian product) of the
+-- output streams of two unfolds.
+--
+{-# INLINE_NORMAL cross #-}
+cross :: Monad m => Unfold m a b -> Unfold m a c -> Unfold m a (b, c)
+cross (Unfold step1 inject1) (Unfold step2 inject2) = Unfold step inject
+
+    where
+
+    inject x = do
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Unfold/Types.hs:115]: Can you rename `x` and `sy` to `seed`?

File: src/Streamly/Internal/Data/Unfold/Types.hs [hunk 4]
```diff
@@ -63,3 +67,86 @@ map f (Unfold ustep uinject) = Unfold step uinject
...
+apDiscardSnd (Unfold _step1 _inject1) (Unfold _step2 _inject2) = undefined
+
+data Cross s1 s2 sy x y = CrossOuter s1 y | CrossInner s1 sy s2 x
+
+-- | Create a cross product (vector product or cartesian product) of the
+-- output streams of two unfolds.
+--
+{-# INLINE_NORMAL cross #-}
+cross :: Monad m => Unfold m a b -> Unfold m a c -> Unfold m a (b, c)
+cross (Unfold step1 inject1) (Unfold step2 inject2) = Unfold step inject
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Unfold/Types.hs:111]: You can maybe rename this to `crossWith` like how we have `teeWith`.

File: src/Streamly/Internal/Data/Unfold/Types.hs (hunk 2 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Unfold/Types.hs:125]: `CrossInner` is a loop, its not a one shot thing.

**Review by adithyaov — COMMENTED** [review]
Can you fix the CI. Can you also add a benchmark for `cross` or open an issue to do so?

File: src/Streamly/Internal/Data/Unfold/Types.hs (hunk 4 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Unfold/Types.hs:111]: This does not accept a function so its more like `tee` than `teeWith`.

**Comment by harendra-kumar** [comment]
> Can you fix the CI. Can you also add a benchmark for `cross` or open an issue to do so?

The `toNullAp` benchmark measures the performance of `cross`.

File: src/Streamly/Internal/Data/Unfold/Types.hs (hunk 3 above)
Comment (harendra-kumar) [reply src/Streamly/Internal/Data/Unfold/Types.hs:115]: ok.

File: src/Streamly/Internal/Data/Unfold/Types.hs (hunk 4 above)
Comment (adithyaov) [reply src/Streamly/Internal/Data/Unfold/Types.hs:111]: Yes, what I meant was to make it accept a function and name it `crossWith`

File: src/Streamly/Internal/Data/Unfold/Types.hs [hunk 5]
```diff
@@ -63,3 +67,86 @@ map f (Unfold ustep uinject) = Unfold step uinject
...
+
+    where
+
+    inject a = do
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Unfold/Types.hs:115]: I like `seed` here as it is shared and is the same in the entire scope of the function.

**Review by adithyaov — APPROVED** [review]

**Comment by harendra-kumar** [comment]
merged offline, commit 3947fee6

**Pull request closed without merging**